torch/csrc/TypeInfo.cpp

```diff
@@ -0,0 +1,284 @@
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject* THPIInfo_pynew(PyTypeObject* type, PyObject* args, PyObject* kwargs) {
+  HANDLE_TH_ERRORS
+  static torch::PythonArgParser parser({
+      "ScalarType(ScalarType type)",
+  });
+  torch::ParsedArgs<1> parsed_args;
+  auto r = parser.parse(args, kwargs, parsed_args);
+  AT_CHECK(r.idx == 0, "Not a type");
+
+  at::ScalarType scalar_type = r.scalartype(0);
+  if (at::isFloatingType(scalar_type)) {
+    return PyErr_Format(
+        PyExc_TypeError,
+        "torch.iinfo() requires an integer input type. Use torch.finfo to handle '%s'",
+        type->tp_name);
+  }
+  return THPIInfo_New(scalar_type);
+  END_HANDLE_TH_ERRORS
+}
+
+static PyObject* THPDTypeInfo_bits(THPDTypeInfo* self, void*) {
+  int bits = elementSize(self->type) * 8;
+  return PyLong_FromLong(bits);
+}
+
+static PyObject* THPFInfo_eps(THPFInfo* self, void*) {
+  switch (self->type) {
+    case at::ScalarType::Float:
+      return PyFloat_FromDouble(std::numeric_limits<float>::epsilon());
+    case at::ScalarType::Double:
+      return PyFloat_FromDouble(std::numeric_limits<double>::epsilon());
+    case at::ScalarType::Half:
+      return PyFloat_FromDouble(std::numeric_limits<at::Half>::epsilon());
+    case at::ScalarType::ComplexFloat:
+      return PyFloat_FromDouble(std::numeric_limits<float>::epsilon());
+    case at::ScalarType::ComplexDouble:
+      return PyFloat_FromDouble(std::numeric_limits<double>::epsilon());
+    case at::ScalarType::ComplexHalf:
+      return PyFloat_FromDouble(std::numeric_limits<at::Half>::epsilon());
+    default:
+      return Py_NotImplemented;
+  }
+}
+
+static PyObject* THPFInfo_max(THPFInfo* self, void*) {
+  switch (self->type) {
+    case at::ScalarType::Float:
+      return PyFloat_FromDouble(std::numeric_limits<float>::max());
+    case at::ScalarType::Double:
+      return PyFloat_FromDouble(std::numeric_limits<double>::max());
+    case at::ScalarType::Half:
+      return PyFloat_FromDouble(std::numeric_limits<at::Half>::max());
+    case at::ScalarType::ComplexFloat:
+      return PyFloat_FromDouble(std::numeric_limits<float>::max());
+    case at::ScalarType::ComplexDouble:
+      return PyFloat_FromDouble(std::numeric_limits<double>::max());
+    case at::ScalarType::ComplexHalf:
+      return PyFloat_FromDouble(std::numeric_limits<at::Half>::max());
+    default:
+      return Py_NotImplemented;
+  }
+}
+
+static PyObject* THPIInfo_max(THPFInfo* self, void*) {
+  switch (self->type) {
+    case at::ScalarType::Byte:
+      return PyLong_FromLong(std::numeric_limits<unsigned char>::max());
+    case at::ScalarType::Char:
+      return PyLong_FromLong(std::numeric_limits<char>::max());
+    case at::ScalarType::Short:
+      return PyLong_FromLong(std::numeric_limits<short>::max());
+      break;
+    case at::ScalarType::Int:
+      return PyLong_FromLong(std::numeric_limits<int>::max());
+    case at::ScalarType::Long:
+      return PyLong_FromLong(std::numeric_limits<long>::max());
```

Comment:
You can probably just use the dispatch macros from `<ATen/Dispatch.h>`. e.g.:

```c++
return AT_DISPATCH_INTEGRAL_TYPES(at::CPU(self->type), "max", [] {
  return THPUtils_packInt64(std::numeric_limits<scalar_t>::max());
});
```

It might be missing a case for floating-point + complex, which you would have to add.

Reply:
Done.